datasync/chaindatafetcher/kafka/kafka.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2020 The klaytn Authors
+// This file is part of the klaytn library.
+//
+// The klaytn library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The klaytn library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the klaytn library. If not, see <http://www.gnu.org/licenses/>.
+
+package kafka
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/Shopify/sarama"
+	"github.com/hashicorp/go-uuid"
+	"github.com/klaytn/klaytn/datasync/chaindatafetcher/types"
+	"github.com/klaytn/klaytn/log"
+)
+
+var (
+	logger = log.NewModuleLogger(log.ChainDataFetcher)
+	kb     *KafkaBroker
+	once   sync.Once
+)
+
+type KafkaBroker struct {
+	producer   sarama.AsyncProducer
+	admin      sarama.ClusterAdmin
+	brokers    []string
+	handlers   map[string]func(*sarama.ConsumerMessage) error
+	consumer   *Consumer
+	replicas   int16
+	partitions int32
+}
+
+func New(groupID string, brokerList []string, replicas int16, partitions int32) types.EventBroker {
+	once.Do(func() {
+		kb = &KafkaBroker{
+			brokers:    brokerList,
+			handlers:   map[string]func(*sarama.ConsumerMessage) error{},
+			replicas:   replicas,
+			partitions: partitions,
+		}
+		kb.newClusterAdmin()
+		kb.newProducer()
+
+		// TODO-ChainDataFetcher context has to be passed by outside.
+		kb.consumer = NewConsumer(context.Background(), kb.newConsumer(groupID))
+	})
+
+	return kb
+}
+
+func (r *KafkaBroker) Publish(topic string, msg interface{}) error {
```

Comment:
`r` for `KafkaBroker` looks confusing.

Reply:
Renamed the receiver to `k`. 1f989f2dcfbb1c9eeed58c88629d79559f246f26